# Ans 1

In [ ]:
print(1.5e200 *2.0e210)


inf


inf = infinity

In [ ]:
print(3.4e-320 / 1e100)

0.0


# Ans 2

In [ ]:
print(1/3)

0.3333333333333333


In [ ]:
print(format(1/3, '.3f'))

0.333


In [ ]:
print(format(1/3, '.1f'))

0.3


In [ ]:
print(6*1/3)

2.0


In [ ]:
print(1/3 + 1/3 + 1/3 + 1/3 + 1/3 + 1/3)

1.9999999999999998


# Ans 3

Difference between a and b is :
In a) value after decimal is printed till maximum value
In b) we have asked to print till 3 decimal points using format specifier

Difference between d and e is:
In d) as per associativity rule 6 will be multiplied 1 and then the resultant number will be divided by 3
But in e) as / and + has different associativity BODMAS rule is applied

# Ans 4

In [ ]:
print(format(22/7, '.5f'))

3.14286


In [ ]:
print("%-10s" % ('Hello', ))

Hello     


In [ ]:
print(format("Hello", "<10"))

Hello     


In [ ]:
print(format("Hello", ">10"))

     Hello


In [ ]:
print(format("Hello", "#^15"))

#####Hello#####


# Ans 5

In [ ]:
print(ord('a'))
print(ord('A'))
print(ord('z'))
print(ord('Z'))

97
65
122
90


In [ ]:
print(ord('0'))
print(ord('9'))
print(ord(' '))
print(ord('\n'))
print(ord('\t'))

48
57
32
10
9


In [ ]:
print(chr(35))
print(chr(67))
print(chr(50))
print(chr(99))

#
C
2
c


For lower case : 97 - 122
For upper case : 65 - 90
For digits : 48 -57

# Ans 6

In [ ]:
print("abc","\n","bcd")

abc 
 bcd


In [ ]:
print("abc\nbcd")

abc
bcd


In [ ]:
print(1,'\n',2)

1 
 2


In [ ]:
print("hello \n\n!!")

hello 

!!


In [ ]:
print("This is '\t' means tab")

This is '	' means tab


4 spaces are printed using \t

In [ ]:
print("This","is","different",sep='\t')

This	is	different


In [ ]:
print("This","is","different",sep='#')

This#is#different


In [ ]:
print("This","is","different",sep='\n')

This
is
different


In [ ]:
print("This","is","different",end='\n')

This is different


In [ ]:
print(10 / 2, 10 // 2.0, 101 / 3, 101 // 3, 101 % 3, 2 ** 4, 3 ** 2, sep="\n")

5.0
5.0
33.666666666666664
33
2
16
9


In [ ]:
abs(-10) # give absolute value

10

In [ ]:
divmod(101,3) # divmod gives two values , interger division and remainder

(33, 2)

In [ ]:
print(8 << 1) # 8 x 2^1

16


In [ ]:
print(8 << 2)

32


In [ ]:
print(32 >> 2) # 32 / 2^2

8


In [ ]:
print(16 >> 1)

8


# Ans 8

In [ ]:
a,b,c = int(input("Enter marks : ").split())
total = int(a) + int(b) + int(c)
print(total)

Q9. Manually calculate output of following expressions using precedence and associativity rules
and then verify output using python shell
a. 2 + 3 * (4 -1)
b. 2 + ((3 *4) - 8)
c. (2 + 3) * 4
d. -10 + 25 / (16 + 12)
e. 2 ** 2 ** 2
f. (3 ** 2) ** 3
g. x = y = 3 + 5
h. 8/2*(2+2)

In [ ]:
print(2 + 3 * (4 -1))

11


In [ ]:
print(2 + ((3 *4) - 8))

6


In [ ]:
print((2 + 3) * 4)

20


In [ ]:
print(-10 + 25 / (16 + 12))

-9.107142857142858


In [ ]:
print(2 ** 2 ** 2)

256


In [ ]:
print((3 ** 2) ** 3)

729


In [ ]:
print(x = y = 3 + 5)

SyntaxError: invalid syntax (2559457631.py, line 1)

In [ ]:
x = y = 3 + 5
print(x)
print(y)


8
8


In [ ]:
print(8/2*(2+2))

16.0


Q10. Check output of following
a. x = y = 10
b. x1 = 20 = y1

In [ ]:
x = y = 10
print(x)
print(y)

10
10


In [ ]:
x1 = 20 = y1

SyntaxError: cannot assign to literal (1868644801.py, line 1)

Q11. What is output of following
a. x=10
y=34.4
print(x and y)
print(x or y)

In [ ]:
x=10
y=34.4
print(x and y)  #if x and y are true then print y else 0
print(x or y)

34.4
10


# IF ELSE

1.	A student will not be allowed to sit in exam if his/her attendence is less than 75%.  
Take following input from user Number of classes held Number of classes attended. And print percentage of class attended  
Is student is allowed to sit in exam or not.

In [ ]:
import math
class_held=int(input("Enter number of classes held : "))
class_att=int(input("Enter number of classes attended : "))
percentage=(class_att/class_held)*100
print("Attendence percentage is : ",math.ceil(percentage), '%') #use math.ceil() to print value of nearest integer
if percentage >= 75:
    print("You are allowed to sit in the exam !!")
else:
    print("You are not allowed to sit in the exam :( ")


Enter number of classes held : 100
Enter number of classes attended : 76
Attendence percentage is :  76 %
You are allowed to sit in the exam !!


2. accept amount from user and find the minimum number notes required to get the amount amount =512  
Notes: 2000,500,100,50,10,5,2,1  

500-1 note  
10  - 1 note  
2-  1 coin  

amount=20550  
2000 – 10 note  
500 – 1 note  
50 -1 note

In [ ]:
amount=int(input("Enter your amount: "))

if (amount>=2000):
    note2k=amount//2000
    print("2000 notes: ", note2k)
    rem=amount%2000
    amount=rem
if (2000>amount>=500):
    note500=amount//500
    print("500 notes: ", note500)
    rem=amount%500
    amount=rem
if (500>amount>=200):
    note200=amount//200
    print("200 notes: ", note200)
    rem=amount%200
    amount=rem
if (200>amount>=100):
    note100=amount//100
    print("100 notes: ", note100)
    rem=amount%100
    amount=rem

if (100>amount>=20):
    note20=amount//20
    print("20 notes: ", note20)
    rem=amount%20
    amount=rem
if (20>amount>=10):
    note10=amount//10
    print("10 notes: ", note10)
    rem=amount%10
    amount=rem
if (10>amount>=5):
    note5=amount//5
    print("5 notes: ", note5)
    rem=amount%5
    amount=rem
if (5>amount>=2):
    note2=amount//2
    print("2 notes: ", note2)
    rem=amount%2
    amount=rem
if (2>amount>=1):
    note1=amount//1
    print("1 notes: ", note1)
    rem=amount%1
    amount=rem



Enter your amount: 5668
2000 notes:  2
500 notes:  3
100 notes:  1
20 notes:  3
5 notes:  1
2 notes:  1
1 notes:  1


In [ ]:
amount=int(input("Enter your amount: "))
arr=[2000,500,100,50,20,10,5,2,1]
for i in arr:
    if (amount>=i):
        note=amount//i
        print("{0} notes: ".format(i), note)
        rem=amount%i
        amount=rem
#    print(i)

Enter your amount: 15423
2000 notes:  7
500 notes:  2
100 notes:  4
20 notes:  1
2 notes:  1
1 notes:  1


3.	Modify the above question Q1 to allow student to sit if he/she has medical cause. Ask user if he/she has medical cause or not ( 'Y' or 'N' ) and print accordingly.

In [ ]:
import math
class_held=int(input("Enter number of classes held : "))
class_att=int(input("Enter number of classes attended : "))
med=input("he/she has medical cause or not? ( 'Y' or 'N' )")
percentage=(class_att/class_held)*100
print("Attendence percentage is : ",math.ceil(percentage), '%') #use math.ceil() to print value of nearest integer
if percentage >= 75:
    print("You are allowed to sit in the exam !!")
elif (med == "Y"):
    print("You are allowed to sit in the exam !!")
else:
    print("You are not allowed to sit in the exam :( ")

Enter number of classes held : 100
Enter number of classes attended : 65
he/she has medical cause or not? ( 'Y' or 'N' )Y
Attendence percentage is :  65 %
You are allowed to sit in the exam !!


4.	A school has following rules for grading system:  
a.	Below 25 - F  
b.	25 to 45 - E  
c.	45 to 50 - D  
d.	50 to 60 - C  
e.	60 to 80 - B  
f.	Above 80 - A  
Ask user to enter marks and print the corresponding grade.

In [ ]:
marks=int(input("Enter your Marks: "))
if marks<25:
    print("F")
elif 25<=marks<45:
    print("E")
elif 45<=marks<50:
    print("D")
elif 50<=marks<60:
    print("C")
elif 60<=marks<80:
    print("B")
elif 80<=marks<=100:
    print("A")
else:
    print("Please enter valid input")


Enter your Marks: 56
C


In [ ]:
marks=int(input("Enter your Marks: "))
arr = [0, 25, 45, 50, 60, 80, 100]
arr_ch = ['F', 'E', 'D', 'C', 'B', 'A']
for i in range(6):
    if arr[i]<=marks<arr[i+1]:
        print(arr_ch[i])

Enter your Marks: 46
D


5.	  If  x = 2 y = 5 z = 0  then find values of the following expressions and Print the output of following statements  
a.	x == 2  
b.	x != 5  
c.	x != 5 & y >= 5  
d.	z != 0 | x == 2  
e.	not(y < 10)  

In [ ]:
x = 2
y = 5
z = 0
print(x == 2)
print(x != 5)
print(x != 5 & y >= 5)
print(z != 0 | x == 2)
print(not(y < 10))

True
True
True
True
False


6.	Accept number from user and check whether it is divisible by 5 and 11 if divisible then display appropriate message.

In [ ]:
num=int(input("Enter a number: "))
if (num%5==0) & (num%11==0):
    print("Divisible by both 5 and 11")
else:
    print("Not divisible")

Enter a number: 14
Not divisible


7.	Write a program to calculate the electricity bill (accept number of unit from user) according to the following criteria :  
             Unit                                    Price    
First 100 units                                       no charge  
Next 100 units                                        Rs 5 per unit  
After 200 units                                       Rs 10 per unit  
(For example if input unit is 350 than total bill amount is Rs2000)  

In [ ]:
unit = int(input("Enter the units : "))
charge = 0
if unit>200:
    charge = charge + (unit-200) * 10
    unit = 200
if 200>=unit>100:
    charge = charge + (unit-100) * 5
    unit = 100
if 100>unit:
    charge = charge + (unit-0) * 0
    unit = 0
print(charge)

Enter the units : 350
2000


8.	Write a program to check whether the last digit of a number( entered by user ) is divisible by 3 or not.  


In [ ]:
num=int(input("Enter a number: "))
digit = num % 10
if digit%3==0:
    print("Number is divisible by 3")
else:
    print("Number is not divisible by 3")

Enter a number: 452
Number is not divisible by 3


9.	Write a program to check whether an years is leap year or not the year is leap if it satisfies following condition  
•	It the year is divisible by 100 o If it is divisible by 100, then it should also be divisible by 400 then it is leap year  
•	otherwise, all other years divisible by 4 and not divisible by 100 then it is leap year.  

In [ ]:
year=int(input("Enter a year: "))
if year % 100 == 0:
    if year % 400 == 0:
        print("Leap year")
    else:
        print("Not a leap year")
elif year % 4 == 0:
    print("Leap year")
else:
    print("Not a leap year")

Enter a year: 2023
Not a leap year


10. Write a program to accept the price of a bike and display the road tax and insurance to be paid according to the following criteria . also display total amount to be paid.  
      
        Cost price (in Rs)           Tax                Inssurance  
        > 100000                     15 %                   20%  
        > 50000 and <= 100000        10%                    8%  
        <= 50000                     5%                     5%   

In [ ]:
price = int(input("Enter price of the bike : "))
total = 0
if price>=100000:
    tax = price * 0.15
    insurance = price * 0.2
    total = price + tax + insurance
elif 100000>price>=50000:
    tax = price * 0.1
    insurance = price * 0.08
    total = price + tax + insurance
elif 50000>=price:
    tax = price * 0.05
    insurance = price * 0.05
    total = price + tax + insurance
print(total)

Enter price of the bike : 25000
27500.0


# Loop

1.	Accept 10 integers from user  and print their average value on the screen 

In [18]:
total=0
avg=0
for i in range(1,11):
    a=int(input("Enter"))
    total=total+a

avg=(total/10)
print("avg: ", avg)

Enter1
Enter5
Enter8
Enter8
Enter9
Enter3
Enter4
Enter3
Enter5
Enter8
avg:  5.4


2.	Print the following patterns using loop :

Print the following patterns using loop : 
a. 
* 
** 
***
****

In [20]:
a=4

for i in range(1,a+1):
    print("*"*i)


*
**
***
****


b. 
    *                    
  ***                
***** 
  ***
    * 

In [36]:
n=5
sp=n//4
for i in range(1,n+1,2):
    print(" "*(n-i),"*"*i,sep="")

for i in range(2,n,2):
    print(" "*(i),"*"*(n-i),sep="")
    

    *
  ***
*****
  ***
    *


In [29]:
num=int(input("Enter an odd number : "))
space=num-1
if num % 2 != 0:
    for i in range(1, num+1, 2):
        print(" "*space, "*"*i, sep="")
        space-=2
    space=2    
    for j in range(num-2, 0, -2):
        print(" "*space, "*"*j, sep="")
        space=space+2
        


Enter an odd number : 9
        *
      ***
    *****
  *******
*********
  *******
    *****
      ***
        *


In [22]:
#program for experimentation purpose
num=int(input("Enter an odd number : "))
space=num-1
start=1
end=num+1
r=2
if num % 2 != 0:
    for i in range(start, end, r):
        print(" "*space, "*"*i, sep="")
        space-=2
        if i==num:
            start, end=end, start
            r=-2
            
            

Enter an odd number : 5
    *
  ***
*****


c. 
1010101          
 10101  
  101   
   1  

In [13]:
n=4
space=0
for i in range(1,n+1):
    print(' '*space,'10'*(n-i),'1',sep='')
    space+=1

1010101
 10101
  101
   1


In [14]:
n=4
a=0
for i in range(1,n+1):
    print(" "*(n-i),"10"*a,"1",sep="")
    a+=1

   1
  101
 10101
1010101


d.
1  
1 2  
1 2 3  
1 2 3 4  
1 2 3 4 5 

In [83]:
n=5

for i in range(1,n+1):
    for j in range (1,i+1):
        print(j,end="")
    print("")

1
12
123
1234
12345


In [1]:
num=int(input("Enter a number : "))
for i in range(1, num+1):
    for j in range(1,i+1):
        print(j, end="")
    print("")

Enter a number : 6
1
12
123
1234
12345
123456


3.	Write a program to find greatest common divisor (GCD) or highest common factor (HCF) of given two numbers.


In [26]:
import math
x=int(input("Enter number 1 : "))
y=int(input("Enter number 2 : "))
print("GCD is the numbers are : ", math.gcd(x,y))

Enter number 1 : 12
Enter number 2 : 108
GCD is the numbers are :  12


4.	Take integer inputs from user until he/she presses q ( Ask to press q to quit after every integer input ). Print average and product of all numbers.

In [6]:
average=0
product=1
x=1
while x:
    x=input("Enter number (press q to exit): ")
    if (x=='q') or (x=='Q'):
        break
    average+=int(x)
    product*=int(x)
    print("Average is: ", average)
    print("Product is: ", product)
    
        

Enter number (press q to exit): 45
Average is:  45
Product is:  45
Enter number (press q to exit): 44
Average is:  89
Product is:  1980
Enter number (press q to exit): q


5.	Given a number count the total number of digits in a number and also find sum of digits of the number.

In [11]:
num=input("Enter a number : ")
total=0
if num.isdigit()==True:
    print("Lenght of the number is :", len(num))
    for i in num:
        total+=int(i)
    print("Total of the digits is :", total)
    

Enter a number : 123456789
Lenght of the number is : 9
Total of the digits is : 45


6.	To display the cube of the number upto given an integer. If the given integer is 5, then display cube of 1 to 4.

In [14]:
num=int(input("Enter a number : "))
for i in range(1,num):
    print(i**3)

Enter a number : 5
1
8
27
64


7.	Accept 20 numbers from user and display sum of only even numbers.

In [ ]:
total=0
for i in range(1,21):
    a=int(input("Enter {0} number: ".format(i)))
    if a%2==0:
        total+=a  
print("Sum of even numbers is: ", total)

8.	Ask user number of terms to be generated of a series. 
generate numbers for the following series and find its addition 
[9 + 99 + 999 + 9999+……..] 


In [2]:
num=int(input("Enter number of terms: "))
total=0
print("Terms are: ", end=" ")
for i in range(1,num+1):
    a= '9'*i
    if i < (num):
        print(a, '+ ', end="")
        total+= int(a)
    else:
        print(a)
        total+= int(a)
print("Total is :", total)

Enter number of terms: 5
Terms are:  9 + 99 + 999 + 9999 + 99999
Total is : 111105


9.	Write a program in python to display the sum of the series [ 1+x+x^2/2!+x^3/3!+....]. Go to the editor 
Test Data : 
Input the value of x :3 Input number of terms : 5 Expected Output : 
The sum is : 16.375000 

In [3]:
import math
x=int(input("Enter value of x : "))
num=int(input("Enter numer of terms : "))
total=0
for i in range(num):
    total+= ((pow(x, i))/math.factorial(i))
print(total)


Enter value of x : 3
Enter numer of terms : 5
16.375


10.	Write a program in python to find the sum of the series [ x - x^3 + x^5 –x^7+x^9-x^11+ ......]. Go to the editor
Test Data :
Input the value of x :2 Input number of terms : 5 Expected Output :
The values of the series:
2
-8
32
-128
512
The sum = 410

In [4]:
import math
x=int(input("Enter value of x : "))
num=int(input("Enter numer of terms : "))
total=0
term=1
for i in range(1,num*2,2):
    if term % 2 !=0:
        total+= (pow(x, i))
    else:
        total-= (pow(x, i))
    term+=1
print(total)


Enter value of x : 2
Enter numer of terms : 5
410


11.	Take a number from user and check if it is prime or not


In [13]:
import math
num=int(input("Enter a number : "))
flag=1
e=int(math.sqrt(num))
for i in range(2,e):
    if num%i==0:
        flag=0
if flag==1:
    print("Number is prime number!")
else:
    print("Number is not a prime :(")

Enter a number : 71
Number is prime number!


12.	Take a number from user and print sum of all odd numbers till that number.
Ex. Enter a no : 10   
Sum of all odd numbers till 10 : 3+5+7+9  =24

In [43]:
num=int(input("enter a number: "))

total=0
for i in range(1,num):
    if (i%2)!=0:
        total+=i
print("Total is: ",total)

enter a number: 10
Total is:  25


# Advance Questions

Q. Create a variable storing integer value. Print the identity of that variable. Then change value of that variable and again print the identity.


In [14]:
var=10
id(var)

140717127406664

In [15]:
var=12
id(var)

140717127406728

Id of var is changed as the value assigned to it is changed

-------------------------------------------------------------------------------

------------------------------------------------------------------------------------------------------

Q. What is PEMDAS rule ? What is BODMAS ? Is there any difference in them ?


<b>PEMDAS</b> rule states that the order of operation starts with the parentheses first or the calculation which is enclosed in brackets. Then the operation is performed on exponents(degree or square roots) and later we do operations on multiplication & division and at last addition and subtraction.



The full form of BODMAS is

Brackets (Numbers inside the Bracket always comes first)

Precedence of brackets are : (bar)

Orders (the power)

Division

Multiplication

Addition

Subtraction

PEMDAS is the proper word to use when solving equations in the United States, while BODMAS is the proper word to use in the United Kingdom.
In BODMAS, division and multiplication have equal priority. In contrast, in PEMDAS, multiplication takes precedence over division3. The order of operations for brackets, orders, addition, subtraction, multiplication, and division is the same for both rules4